### PR TITLE
Dont push too often to Octopus Cloud

### PR DESCRIPTION
**Pull request by matt-richardson**
Upgrades on OctopusCloud at the moment take the instance down, causing a customer outage - we dont want to cause an outage every day. 

We want to remove this check as soon as we've got 0 downtime upgrades for Octopus Cloud instances, which will hopefully be cycle 2 or cycle 3 2021.

### CODE REVIEW

File: enthusiastic-promoter.ps1 [hunk 1]
```diff
@@ -1,15 +1,15 @@
...
+    "Environments-2584" = @{ "Name" = "Branch Instances (Prod)";    "BakeTime" = New-TimeSpan -Days 1;    "StabilizationPhaseBakeTime" = New-TimeSpan -Days 1;    "MinimumTimeBetweenDeployments" = New-TimeSpan -Minutes 0; }
+    "Environments-2585" = @{ "Name" = "Staff";                      "BakeTime" = New-TimeSpan -Minutes 0; "StabilizationPhaseBakeTime" = New-TimeSpan -Days 1;    "MinimumTimeBetweenDeployments" = New-TimeSpan -Minutes 0; }
+    "Environments-2586" = @{ "Name" = "Friends of Octopus";         "BakeTime" = New-TimeSpan -Minutes 0; "StabilizationPhaseBakeTime" = New-TimeSpan -Days 1;    "MinimumTimeBetweenDeployments" = New-TimeSpan -Hours 12; }
+    "Environments-2587" = @{ "Name" = "Early Adopters";             "BakeTime" = New-TimeSpan -Minutes 0; "StabilizationPhaseBakeTime" = New-TimeSpan -Days 7;    "MinimumTimeBetweenDeployments" = New-TimeSpan -Days 3; }
+    "Environments-2588" = @{ "Name" = "Stable";                     "BakeTime" = New-TimeSpan -Minutes 0; "StabilizationPhaseBakeTime" = New-TimeSpan -Days 7;    "MinimumTimeBetweenDeployments" = New-TimeSpan -Days 7; }
```
Comment (matt-richardson) [inline enthusiastic-promoter.ps1:11]: Happy for better suggestions for these numbers (either now, or we adjust on the fly)

File: enthusiastic-promoter.ps1 [hunk 2]
```diff
@@ -84,6 +84,7 @@ function Test-PipelineBlocked($release) {
 }
 
 function Get-CurrentEnvironment($progression, $release) {
+    if ($release.NextDeployments.Length -eq 0) { return $null }
```
Comment (matt-richardson) [inline enthusiastic-promoter.ps1:87]: This line comes from [this pr](https://github.com/OctopusDeploy/OctopusDeploy-EnthusiasticPromotions/pull/13/files#diff-ba11940861ac3f1f9c4f4551695e1fd2cc275d7f76706aeccb8c5832fa2dfe54R92) - it was causing errors

File: enthusiastic-promoter.ps1 [hunk 3]
```diff
@@ -229,6 +241,10 @@ function Get-PromotionCandidates($progression, $channels, $lifecycles) {
             } elseif (($null -ne $mostRecentReleaseDeployedToNextEnvironment) -and ((New-Object Octopus.Versioning.Semver.SemanticVersion $mostRecentReleaseDeployedToNextEnvironment.Release.Version) -gt (New-Object Octopus.Versioning.Semver.SemanticVersion $release.Release.Version))) {
                 $channelName = Get-ChannelName $channels $release.Release.ChannelId
                 Write-Host " - A newer release '$($mostRecentReleaseDeployedToNextEnvironment.Release.Version)' in channel '$channelName' has already been deployed to '$nextEnvironmentName'."
+            } elseif (Test-ShouldLimitDeploymentsToEnvironment -nextEnvironmentId $nextEnvironmentId -mostRecentReleaseDeployedToNextEnvironment $mostRecentReleaseDeployedToNextEnvironment) {
+                $minimumTimeBetweenDeployments = $waitTimeForEnvironmentLookup[$nextEnvironmentId].MinimumTimeBetweenDeployments
+                $mostRecentDeploymentToNextEnvironment = Get-MostRecentDeploymentToEnvironment $mostRecentReleaseDeployedToNextEnvironment $nextEnvironmentId
```
Comment (matt-richardson) [inline enthusiastic-promoter.ps1:246]: This is duplicated between line 203/204. Trade off between "have one copy of the logic" vs "keep the `Get-PromotionCandidates` function responsible for the logging". Not sure I landed on the right side of the fence there.

File: enthusiastic-promotor.tests.ps1 [hunk 4]
```diff
@@ -33,31 +33,28 @@ Describe 'Enthusiastic promoter' {
...
+    $result[2].ChannelName | Should -be "CI Builds"
   }
 
-  It 'should only promote 2020.4.7' {
-    # everything else is either:
-    # * baking
-    # * progressed as far as it can
-    # * had a deployment attempted, but it failed
-    # * had a deployment attempted - its still executing or queued
+  It 'scenario 2' {
```
Comment (Silcoish) [inline enthusiastic-promotor.tests.ps1:51]: Suggest giving a better name, maybe something like 'should not promote anything as all releases are blocked`

**Review by Silcoish — APPROVED** [review]
Code changes look good, I'm happy with the code duplication you mentioned in your comment, keeping the logging together is good to me. I'll leave it to @pawelpabich to comment on the wait times between deployment to environments

File: enthusiastic-promoter.ps1 (hunk 3 above)
Comment (pawelpabich) [reply enthusiastic-promoter.ps1:251]: Idea:  If nested ifs could be replaced by early returns using `continue` then 
  $minimumTimeBetweenDeployments = $waitTimeForEnvironmentLookup[$nextEnvironmentId].MinimumTimeBetweenDeployments
                $mostRecentDeploymentToNextEnvironment = Get-MostRecentDeploymentToEnvironment $mostRecentReleaseDeployedToNextEnvironment $nextEnvironmentId``
`
could be executed before `Test-ShouldLimitDeploymentsToEnvironment` which now would be only responsible for performing the date/time comparision

**Review by pawelpabich — APPROVED** [review]
Didn't check the tests but the rest looks ok to me.

File: enthusiastic-promoter.ps1 (hunk 3 above)
Comment (matt-richardson) [reply enthusiastic-promoter.ps1]: I've refactored into a separate method - was planning on doing that at some point as it was getting hairy anyway 👍

**Pull request merged**